File: crates/prova-core/testdata/shell_argv.lua
-- The argv form of `shell.run` / `shell.spawn`: no shell, no quoting — mirroring `container:run`,
-- whose docs sell exactly that. Its absence was the local half of the SDK lacking what the
-- containerized half had. See docs/design/agent-ergonomics.md §1.

prova.test("an argv table runs the program directly, preserving arguments verbatim", function(t)
  -- A shell would collapse the run of spaces; argv hands the argument over untouched.
  local r = shell.run({ "echo", "hello   world" })
  t:expect(r.code):equals(0)
  t:expect(r.stdout):equals("hello   world\n")
end)

prova.test("argv passes content a shell string would EXECUTE, as data", function(t)
  -- The whole point: this is what you get when a test passes SQL/source/JSON to a local CLI.
  -- Through a shell string, `;` starts a command and `$(…)` substitutes. Through argv, neither
  -- can happen — there is no shell to interpret them.
  local nasty = [[a "b" ; echo INJECTED $(echo SUBSTITUTED)]]
  local r = shell.run({ "echo", nasty })
  t:expect(r.stdout):equals(nasty .. "\n")
  t:expect(r.stdout):never():contains("INJECTED\n")
end)

prova.test("a string command still goes through a shell", function(t)
  -- The string form is unchanged and still ergonomic for fixed commands.
  local r = shell.run("echo one && echo two")
  t:expect(r.stdout):equals("one\ntwo\n")
end)

prova.test("shell.spawn takes argv too", function(t)
  local proc = t:manage(shell.spawn({ "sleep", "30" }))
  t:expect(proc.pid):is_truthy()
end)

prova.test("a non-string, non-table command fails with an actionable message", function(t)
  -- Previously an argv attempt died as "bad argument #1: error converting Lua table to String",
  -- which names the conversion, not the fix.
  local ok, err = pcall(function() return shell.run(42) end)
  t:expect(ok):is_false()
  t:expect(tostring(err)):contains("argv table")
end)

File: crates/prova-core/tests/shell_argv.rs
use std::path::PathBuf;

use prova_core::{run_path, NullReporter};

/// `shell.run`/`shell.spawn` accept an **argv table** (no shell, no quoting) as well as a shell
/// string — closing the asymmetry with `container:run`, which had argv all along. The injection
/// case is the one that matters: content passed through argv is data, never commands.
/// See docs/design/agent-ergonomics.md §1.
#[cfg_attr(windows, ignore = "the Lua suite drives unix tools (echo/sleep binaries, \n endings)")]
#[test]
fn shell_accepts_argv_as_well_as_a_shell_string() {
    let path = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("testdata")
        .join("shell_argv.lua");
    let mut reporter = NullReporter;
    let summary = run_path(&path, &mut reporter).expect("run");
    assert_eq!(summary.failed, 0, "argv suite had failures: {summary:?}");
    assert!(
        summary.passed >= 5,
        "expected the whole argv suite to run, got {summary:?}"
    );
}
